net/softethervpn/patches/120-openssl-deprecated.patch

```diff
@@ -0,0 +1,210 @@
++	t1 = X509_getm_notBefore(x509);
++	t2 = X509_getm_notAfter(x509);
+ 	if (!UINT64ToAsn1Time(t1, notBefore))
+ 	{
+ 		FreeX509(x509);
+@@ -2050,8 +2060,8 @@ X509 *NewRootX509(K *pub, K *priv, NAME *name, UINT days, X_SERIAL *serial)
+ 	X509_set_version(x509, 2L);
+ 
+ 	// Set the Expiration
+-	t1 = X509_get_notBefore(x509);
+-	t2 = X509_get_notAfter(x509);
++	t1 = X509_getm_notBefore(x509);
++	t2 = X509_getm_notAfter(x509);
+ 	if (!UINT64ToAsn1Time(t1, notBefore))
+ 	{
+ 		FreeX509(x509);
+@@ -2677,6 +2687,43 @@ bool RsaCheckEx()
+ 
+ 	return false;
+ }
++
++// RSA key generation
++static RSA *RsaGenKey(UINT bit, BN_ULONG e)
++{
++	RSA *rsa = NULL;
++	char errbuf[MAX_SIZE];
++	BIGNUM *bne = NULL;
++
++	if ((bne = BN_new()) == NULL)
++	{
++		Debug("BN_new: err=%s\n", ERR_error_string(ERR_get_error(), errbuf));
++		return NULL;
++	}
++	if (BN_set_word(bne, e) == 0)
++	{
++		Debug("BN_set_word: err=%s\n", ERR_error_string(ERR_get_error(), errbuf));
++		goto fail;
++	}
++	if ((rsa = RSA_new()) == NULL)
++	{
++		Debug("RSA_new: err=%s\n", ERR_error_string(ERR_get_error(), errbuf));
++		goto fail;
++	}
++	if (RSA_generate_key_ex(rsa, bit, bne, NULL) == 0)
++	{
++		Debug("RSA_generate_key_ex: err=%s\n", ERR_error_string(ERR_get_error(), errbuf));
++		goto fail;
++	}
++	BN_free(bne);
++	return rsa;
++
++fail:
++	RSA_free(rsa);
++	BN_free(bne);
++	return NULL;
++}
++
+ bool RsaCheck()
+ {
+ 	RSA *rsa;
+@@ -2690,12 +2737,11 @@ bool RsaCheck()
+ 	// Key generation
+ 	Lock(openssl_lock);
+ 	{
+-		rsa = RSA_generate_key(bit, RSA_F4, NULL, NULL);
++		rsa = RsaGenKey(bit, RSA_F4);
+ 	}
+ 	Unlock(openssl_lock);
+ 	if (rsa == NULL)
+ 	{
+-		Debug("RSA_generate_key: err=%s\n", ERR_error_string(ERR_get_error(), errbuf));
+ 		return false;
+ 	}
+ 
+@@ -2760,12 +2806,11 @@ bool RsaGen(K **priv, K **pub, UINT bit)
+ 	// Key generation
+ 	Lock(openssl_lock);
+ 	{
+-		rsa = RSA_generate_key(bit, RSA_F4, NULL, NULL);
++		rsa = RsaGenKey(bit, RSA_F4);
+ 	}
+ 	Unlock(openssl_lock);
+ 	if (rsa == NULL)
+ 	{
+-		Debug("RSA_generate_key: err=%s\n", ERR_error_string(ERR_get_error(), errbuf));
+ 		return false;
+ 	}
+ 
+@@ -3875,7 +3920,7 @@ X *X509ToX(X509 *x509)
+ 				{
+ 					if (OBJ_obj2nid(ad->method) == NID_ad_ca_issuers && ad->location->type == GEN_URI)
+ 					{
+-						char *uri = (char *)ASN1_STRING_data(ad->location->d.uniformResourceIdentifier);
++						char *uri = (char *)ASN1_STRING_get0_data(ad->location->d.uniformResourceIdentifier);
+ 
+ 						if (IsEmptyStr(uri) == false)
+ 						{
+@@ -4088,7 +4133,9 @@ void Rand(void *buf, UINT size)
+ // Delete a thread-specific information that OpenSSL has holded
+ void FreeOpenSSLThreadState()
+ {
++#if OPENSSL_VERSION_NUMBER < 0x10100000L
+ 	ERR_remove_state(0);
++#endif
+ }
+ 
+ // Release the Crypt library
+@@ -4110,13 +4157,14 @@ void InitCryptLibrary()
+ 	CheckIfIntelAesNiSupportedInit();
+ //	RAND_Init_For_SoftEther()
+ 	openssl_lock = NewLock();
++#if OPENSSL_VERSION_NUMBER < 0x10100000L

```

Comment:
```suggestion
+#if OPENSSL_VERSION_NUMBER >= 0x10100000L
+  	OPENSSL_init_crypto(OPENSSL_INIT_LOAD_CONFIG, NULL);
+#else
```
I would change the initialization to load `openssl.cnf` and enable engines, to use hw-crypto, for example.  Note that `SSL_library_init()` is actually defined as `OPENSSL_init_ssl(0, NULL)`, which initializes both ssl and cyrpto libs, and sets `OPENSSL_INIT_LOAD_CONFIG`. This behaviour was started in 1.1.1; in 1.1.0, both `init`s were the same, so I would have left the initialization out, as I had many times when porting code in the past. I don't know the code well enough to know if a libssl function is called before a libcrypto one, which would make this moot.  You may use `OPENSSL_init_ssl(0, NULL)` if you want to keep strict adherence to the original code.  Here's the quote from `OPENSSL_init_crypto()` help documenting this:

>**OPENSSL_INIT_LOAD_CONFIG**
>           With this option an OpenSSL configuration file will be automatically loaded and used by calling OPENSSL_config().
>           This is not a default option for libcrypto.  As of OpenSSL 1.1.1 this is a default option for libssl (see OPENSSL_init_ssl(3) for further details about libssl initialisation).

Reply:
Either I'm tired or it sounds like this is already default. Note that this patch cannot be upstreamed as is since it breaks compatibility with OpenSSL 1.0.2. Not an issue here since 1.1.1 is used.

Reply:
That quote is from the OpenSSL man pages. Loading config is default for init_ssl but not for init_crypto.